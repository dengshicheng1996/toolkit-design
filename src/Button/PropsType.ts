import { BasePropsType, IconType } from '../_internal'

export interface ButtonPropsType extends BasePropsType {
  type?: 'default' | 'primary' | 'ghost' | 'warn' | 'warn-ghost' | 'light'

  disabled?: boolean
  onPress?: (e?: React.SyntheticEvent) => void
  activeClassName?: string

  subTitle?: string
  capsule?: boolean
  capsuleSize?: 'md' | 'lg' | 'sm'
  capsuleAutoWidth?: boolean

  icon?: IconType
  loading?: boolean
  loadingText?: string

  htmlType?: 'button' | 'submit' | 'reset'
}
