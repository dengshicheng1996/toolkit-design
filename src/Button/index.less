@import (reference) '../style/index.less';

// start

/* size | 按钮内间距 */
@button-padding: @size-3;

/* corner-radius| 按钮圆角 */
@button-corner-radius: @corner-radius-md;

/* font-size | 副标题字号 */
@button-subtitle-size: @font-size-subcontent;

/* color | 按钮边框宽度 */
@button-border-size: @border-width-standard;

/* opacity | disable状态不透明度 */
@button-disabled-opacity: @opacity-disabled;

/* color | 默认按钮文字色 */
@button-color: @color-fill-grey-dark;

/* color | 默认按钮背景色 */
@button-background: @color-fill-grey-inverse;

/* color | 默认按钮边框色 */
@button-border-color: @color-fill-grey-light;

/* color |primary按钮背景色 */
@button-primary-background: @color-fill-primary;

/* color |primary按钮边框色 */
@button-primary-border-color: @color-fill-primary;

/* color |primary按钮文字色 */
@button-primary-color: @color-text-base;

/* color |ghost按钮边框色 */
@button-ghost-border-color: @color-fill-primary;

/* color |ghost按钮文字色 */
@button-ghost-color: @color-text-primary;

/* color |warn按钮背景色 */
@button-warn-background: @color-fill-warning;

/* color |warn按钮边框色 */
@button-warn-border-color: @color-fill-warning;

/* color |warn按钮文字色 */
@button-warn-color: @color-text-base;

/* color |warn-ghost按钮边框色 */
@button-warn-ghost-border-color: @color-fill-warning;

/* color |warn-ghost按钮文字色 */
@button-warn-ghost-color: @color-text-warning;

/* color |light按钮背景色 */
@button-light-background: @color-fill-primary-light;

/* color |light按钮边框色 */
@button-light-border-color: @color-fill-primary-light;

/* color |light按钮文字色 */
@button-light-color: @color-text-primary;

/* color |text按钮文字色 */
@button-text-color: @color-fill-primary;

/* size | 胶囊按钮圆角 */
@button-capsule-corner: @corner-radius-md;

/* size | 胶囊按钮水平内间距 */
@button-capsule-padding-h: @size-3;

/* size | 胶囊按钮垂直内间距 */
@button-capsule-padding-v: @size-1;

/* font-size | 胶囊按钮大字号 */
@button-capsule-size-l: @font-size-subtitle;

/* font-size | 胶囊按钮中字号 */
@button-capsule-size-m: @font-size-content;

/* font-size | 胶囊按钮小字号 */
@button-capsule-size-s: @font-size-subcontent;

// end

@prefixCls: tdm-button;

@subtitle-font-opacity: 0.6;

.@{prefixCls} {
  position: relative;
  display: flex;
  height: auto;
  box-sizing: border-box;
  align-items: center;
  justify-content: center;
  padding: @button-padding;
  border: @button-border-size solid @button-border-color;
  background-color: @button-background;
  border-radius: @button-corner-radius;
  color: @button-color;
  font-size: @font-size-title;
  line-height: @default-line-height;
  text-align: center;
  user-select: none;

  &&-disabled {
    background-color: @button-background;
    color: @button-color;
    opacity: @button-disabled-opacity;
  }

  &-active {
    // border-color: lighten(@button-border-color, 20%);
    filter: brightness(0.9);

    > :nth-child(1n + 1) {
      opacity: @opacity-active;
    }
  }

  &-subtitle {
    @font-size-subtitle: 34 * @rpx;

    font-size: @font-size-subtitle;

    &-subfont {
      font-size: @button-subtitle-size;
      opacity: @subtitle-font-opacity;
    }
  }

  &-primary,
  &-primary.@{prefixCls}-disabled {
    border-color: @button-primary-border-color;
    background-color: @button-primary-background;
    color: @button-primary-color;

    .@{prefixCls}-loading {
      // loading svg
      circle {
        stroke: @button-primary-color;
      }
    }
  }

  &-ghost,
  &-ghost.@{prefixCls}-disabled {
    border-color: @button-ghost-border-color;
    background-color: transparent;
    color: @button-ghost-color;
  }

  &-ghost&-active {
    border-color: lighten(@button-ghost-border-color, 20%);
  }

  &-warn,
  &-warn.@{prefixCls}-disabled {
    border-color: @button-warn-border-color;
    background-color: @button-warn-background;
    color: @button-warn-color;

    .@{prefixCls}-loading {
      // loading svg
      circle {
        stroke: @button-warn-color;
      }
    }
  }

  &-warn-ghost,
  &-warn-ghost.@{prefixCls}-disabled {
    border-color: @button-warn-ghost-border-color;
    background-color: transparent;
    color: @button-warn-ghost-color;
  }

  &-warn-ghost&-active {
    border-color: lighten(@button-warn-ghost-border-color, 20%);
  }

  &-light,
  &-light.@{prefixCls}-disabled {
    border-color: @button-light-border-color;
    background-color: @button-light-background;
    color: @button-light-color;
  }

  &-capsule {
    display: inline-flex;
    padding: @button-capsule-padding-v @button-capsule-padding-h;
    border-radius: @button-capsule-corner;

    &-lg {
      min-width: 168 * @rpx;
      font-size: @button-capsule-size-l;
    }

    &-md {
      min-width: 152 * @rpx;
      font-size: @button-capsule-size-m;
    }

    &-sm {
      min-width: 144 * @rpx;
      font-size: @button-capsule-size-s;
    }

    &-auto-width {
      min-width: 0;
    }

    .@{prefixCls}-loading-container {
      display: none;
    }
  }

  &-loading-container {
    position: absolute;
    display: flex;
    align-items: center;
    justify-content: center;
  }

  &-loading {
    // 为了保证线条的宽度
    width: 80 * @rpx;
    height: 80 * @rpx;
    margin-right: 16 * @rpx;

    // loading svg
    circle {
      stroke: @color-fill-primary;
    }
  }

  &-icon {
    &-inner {
      margin-right: 10 * @rpx;
    }

    &-self {
      margin-right: 10 * @rpx;
    }
  }
}
