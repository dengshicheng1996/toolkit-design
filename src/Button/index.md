---
title: 按钮
nav:
  title: 组件
  path: /components
---

### 展示

<code src="./demo/basic.tsx" />


### 参数

| 属性 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| type | 按钮类型 | 'default'\|'primary'\|'ghost'\|'warn' | 'default' |
| disabled | 禁用 | boolean | false |
| onPress | 点击回调 | (e?: React.SyntheticEvent) => void | - |
| className | 类名 | string | '' |
| activeClassName | 激活时类名 | string | '' |
| subTitle | 副文案 | string | '' |
| capsule | 是否胶囊按钮 | boolean | false |
| capsuleSize | 胶囊按钮尺寸 | 'sm'\|'md'\|'lg' | 'md' |
| capsuleAutoWidth | 胶囊按钮默认有个最小宽度，如果设置此属性，则不会有 | boolean | false |
| icon | 图标 | React.ReactNode\|Icon | - |
| loading | 加载状态 | boolean | false |
| loadingText | 加载文字 | string | 'Loading' |
