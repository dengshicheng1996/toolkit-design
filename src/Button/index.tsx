import * as React from 'react';
import classnames from 'classnames';
import { Touchable, withError } from '../rmc';
import { renderIcon, getDataAttr } from '../_internal';
import { ButtonPropsType } from './PropsType';
import LoadingIcon from '../Loading/icon';
// import prefixCls from '../prefixCls'
import './index.less';
const prefixCls = 'tdm-button';
const hiddenStyle = {
  width: 0,
  height: 0,
  overflow: 'hidden',
  padding: 0,
  margin: 0,
  border: 0,
};

export const Button: React.FC<ButtonPropsType> = (props) => {
  const {
    type,
    className,
    activeClassName,
    disabled,
    subTitle,
    capsule,
    capsuleSize,
    capsuleAutoWidth,
    onPress,
    icon,
    loading,
    loadingText,
    htmlType,
  } = props;

  const buttonRef = React.useRef<any>();

  const [cls, setCls] = React.useState({
    [`${prefixCls}`]: true,
    [`${prefixCls}-active`]: false,
  });

  const clsStr = classnames({
    ...cls,
    [`${prefixCls}-${type}`]: true,
    [`${className}`]: !!className,
    [`${prefixCls}-capsule ${prefixCls}-capsule-${capsuleSize}`]: capsule,
    [`${prefixCls}-capsule-auto-width`]: capsuleAutoWidth,
    [`${prefixCls}-disabled`]: disabled,
    [`${prefixCls}-subtitle`]: !!subTitle,
    [`${prefixCls}-icon`]: !!icon,
  });

  return (
    <Touchable
      onPressIn={() => {
        setCls({
          ...cls,
          [`${prefixCls}-active`]: true,
          [activeClassName!]: true,
        });
      }}
      onPressOut={() => {
        setCls({
          ...cls,
          [`${prefixCls}-active`]: false,
          [activeClassName!]: false,
        });
      }}
      onPress={(e) => {
        onPress && onPress(e);
        // trigger button default as submit, button, reset
        buttonRef.current?.click();
      }}
      disabled={disabled}
    >
      <div
        {...getDataAttr(props)}
        className={clsStr}
        role="button"
        aria-disabled={disabled}
      >
        {loading && (
          <div className={`${prefixCls}-loading-container`}>
            <div className={`${prefixCls}-loading`}>
              <LoadingIcon />
            </div>
            <div className={`${prefixCls}-loading-text`}>{loadingText}</div>
          </div>
        )}
        {!loading &&
          renderIcon(icon, {
            nodeClassName: `${prefixCls}-icon-self`,
          })}
        <span style={{ opacity: loading && !capsule ? 0 : 1 }}>
          {props.children}
          {subTitle && (
            <div className={`${prefixCls}-subtitle-subfont`}>{subTitle}</div>
          )}
        </span>
        {htmlType && (
          <button
            ref={(el) => (buttonRef.current = el)}
            onClick={(e) => e.stopPropagation()}
            style={hiddenStyle}
            type={htmlType}
          />
        )}
      </div>
    </Touchable>
  );
};

Button.displayName = 'Button';
Button.defaultProps = {
  className: '',
  activeClassName: '',
  type: 'default',
  disabled: false,
  capsule: false,
  subTitle: '',
  capsuleSize: 'md',
  capsuleAutoWidth: false,
  loading: false,
};

export default withError(Button);
