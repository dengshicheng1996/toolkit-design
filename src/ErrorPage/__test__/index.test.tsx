import * as React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'
import Button from '../../Button'
import { ErrorPage } from '..'

describe('ErrorPage', () => {
  it('ErrorPage with src/text/subText/button snapshot', () => {
    const component = shallow(
      <ErrorPage
        src={
          'https://gw.alipayobjects.com/zos/basement_prod/df4e7c06-88db-43dd-b171-725bd1cbaa33.svg'
        }
        text={'付款没成功'}
        subText={'请重新付款'}
        data-y="111"
      >
        <Button type="ghost" capsuleSize="lg" capsule>
          返回
        </Button>
      </ErrorPage>,
    )

    expect(toJSON(component)).toMatchSnapshot()
  })

  it('ErrorPage with src/text/button snapshot', () => {
    const component = shallow(
      <ErrorPage
        src={
          'https://gw.alipayobjects.com/zos/basement_prod/df4e7c06-88db-43dd-b171-725bd1cbaa33.svg'
        }
        text={'付款没成功'}
      >
        <Button type="ghost" capsuleSize="lg" capsule>
          返回
        </Button>
      </ErrorPage>,
    )

    expect(toJSON(component)).toMatchSnapshot()
  })

  it('ErrorPage with src/text snapshot', () => {
    const component = shallow(
      <ErrorPage
        src={
          'https://gw.alipayobjects.com/zos/basement_prod/df4e7c06-88db-43dd-b171-725bd1cbaa33.svg'
        }
        text={'付款没成功'}
      />,
    )

    expect(toJSON(component)).toMatchSnapshot()
  })
})
