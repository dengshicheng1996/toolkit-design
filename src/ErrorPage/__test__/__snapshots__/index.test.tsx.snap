// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorPage ErrorPage with src/text snapshot 1`] = `
<div
  className="amd-error-page"
>
  <img
    className="amd-error-page-poster"
    src="https://gw.alipayobjects.com/zos/basement_prod/df4e7c06-88db-43dd-b171-725bd1cbaa33.svg"
  />
  <div
    className="amd-error-page-text"
  >
    付款没成功
  </div>
  <div
    className="amd-error-page-sub-text"
  />
  <div
    className="amd-error-page-content"
  />
</div>
`;

exports[`ErrorPage ErrorPage with src/text/button snapshot 1`] = `
<div
  className="amd-error-page"
>
  <img
    className="amd-error-page-poster"
    src="https://gw.alipayobjects.com/zos/basement_prod/df4e7c06-88db-43dd-b171-725bd1cbaa33.svg"
  />
  <div
    className="amd-error-page-text"
  >
    付款没成功
  </div>
  <div
    className="amd-error-page-sub-text"
  />
  <div
    className="amd-error-page-content"
  >
    <Button
      capsule={true}
      capsuleSize="lg"
      type="ghost"
    >
      返回
    </Button>
  </div>
</div>
`;

exports[`ErrorPage ErrorPage with src/text/subText/button snapshot 1`] = `
<div
  className="amd-error-page"
  data-y="111"
>
  <img
    className="amd-error-page-poster"
    src="https://gw.alipayobjects.com/zos/basement_prod/df4e7c06-88db-43dd-b171-725bd1cbaa33.svg"
  />
  <div
    className="amd-error-page-text"
  >
    付款没成功
  </div>
  <div
    className="amd-error-page-sub-text"
  >
    请重新付款
  </div>
  <div
    className="amd-error-page-content"
  >
    <Button
      capsule={true}
      capsuleSize="lg"
      type="ghost"
    >
      返回
    </Button>
  </div>
</div>
`;
